.circleci/parse_yaml.py

```diff
@@ -0,0 +1,273 @@
+
+def load_yaml(yaml_file):
+    """Read the found yaml file and return the read object."""
+    with open(yaml_file, 'r') as args:
+        try:
+            return yaml.safe_load(args)
+        except yaml.YAMLError as exc:
+            logging.error(f' {exc}')
+            sys.exit(1)
+
+
+def verify_weblink(path):
+    """Verify the provided link to the contributed file."""
+    import urllib
+
+    weburl = urllib.request.urlopen(path)
+
+    if weburl.getcode() == 200:
+        return True
+    else:
+        return False
+
+
+def download_file(path, outdir='./'):
+    """Download data based on the provided link.
+
+    Note;
+    Progress bars added based on the following reference:
+    https://stackoverflow.com/questions/37573483/progress-bar-while-download-file-over-http-with-requests/37573701
+    """
+    from tqdm import tqdm
+    import requests
+    import math
+    
+    r = requests.get(path, stream=True)
+    total_size = int(r.headers.get('content-length', 0))
+    block_size = 1024
+    wrote = 0
+    file_name = path.split('/')[-1]
+    file_path = outdir + file_name
+    
+    with open(file_path, 'wb') as f:
+        for data in tqdm(r.iter_content(block_size), 
+                         total=math.ceil(total_size//block_size), 
+                         unit='KB', 
+                         leave=True,
+                         ncols=42,
+                         unit_scale=True,
+                         unit_divisor=1024):
+            wrote = wrote + len(data)
+            f.write(data)
+    
+    if total_size != 0 and wrote != total_size:
+        logging.error(f' check the file link and try again.')
+        sys.exit(1)
+
+
+def validate_hash(path, hash):
+    """Validate a file against an MD5 hash value."""
+    import hashlib
+
+    md5 = hashlib.md5()
+    
+    with open(path, 'rb') as f:
+        while True:
+            chunk = f.read(10000000)
+            if not chunk:
+                break
+            md5.update(chunk)
+    
+    return md5.hexdigest() == hash
+
+
+class Yamler:
+    """A collection of utilities to parse the yaml file, check metadata
+    and trigger cimr processing of the contributed file
+    """
+
+    def __init__(self, yaml_data):
+        self.yaml_data = yaml_data
+        self.data_type = None
+        self.keys = None
+        self.hash = None
+        self.outdir = None
+        self.downloaded_file = None
+
+
+    def pick_keys(self):
+        """List keys for the dictionarized yaml data and store in self.
+        The following keys are expected:
+        ['defined_as', 'data_file', 'contributor', 'data_info', 'method']
+        """
+        self.keys = self.yaml_data.keys()
+
+
+    def set_data_type(self):
+        """One of the following data_type variables are expected:
+        ['gwas', 'twas', 'eqtl', 'sqtl', 'pqtl', 'tad']
+        """
+        if self.yaml_data['data_info']['data_type'] is not None:
+            self.data_type = self.yaml_data['data_info']['data_type']
```

Comment:
Do you need to validate `data_type` here? I know that `cimr` is able to validate data types, but it is probably better to report the error earlier so we don't have to download/decompress an invalid file.

Reply:
What if it is a bulk dowload, in which the tarball file includes multiple folders of different data types, such as:
```
cimr-d_test_data/
├── gwas
│   ├── Coronary_Artery_Disease.txt.gz
│   ├── HDL_Cholesterol_NMR.txt.gz
│   ├── HDL_Cholesterol.txt.gz
│   ├── Heart_Attack_UKB.txt.gz
│   ├── LDL_Cholesterol_NMR.txt.gz
│   └── LDL_Cholesterol.txt.gz
└── tad
    ├── GSE105381.homo_sapiens.HepG2.tad.txt
    ├── GSE105491.homo_sapiens.SK-MEL-5.tad.txt
     ...
    ├── GSE87112.homo_sapiens.Pancreas.tad.txt
    └── GSE87112.homo_sapiens.Spleen.tad.txt
```

Reply:
Currently, nothing. I have now included 'multiple' as a potential option of a data_type to handle something like this. e.g. if data_type == 'multiple': search for dir tree instead of removing all dirs by default from a tar.

Reply:
I have created [a new issue](https://github.com/greenelab/cimr-d/issues/10) as a reminder to address this.